procbuild/listener.py

```diff
@@ -0,0 +1,186 @@
+        self.prefix = prefix
+
+        self.socket = self.ctx.socket(zmq.SUB)
+        self.socket.connect(OUT)
+        self.socket.setsockopt(zmq.SUBSCRIBE, self.prefix.encode('utf-8'))
+        
+        self.queue = asyncio.Queue()
+        self.dont_build = set()
+
+    async def listen(self):
+        """Listener method, containing while loop for checking socket
+        """
+        while True:
+            msg = await self.socket.recv_multipart()
+            target, raw_payload = msg
+            payload = json.loads(raw_payload.decode('utf-8'))
+            print('received', payload)
+            paper_to_build = payload.get('build_paper', None)
+            
+            if self.check_age_and_queue(paper_to_build):
+                continue
+            self.dont_build.add(paper_to_build)
+            await self.queue.put(paper_to_build)
+    
+    def check_age(self, nr):
+        """Check the age of a PR's status_file based on its number. 
+        
+        Parameters:
+        ------------
+        nr: int, the number of the PR in order of receipt
+        """
+        age = file_age(status_file(nr))
+        min_wait = 0.5
+        too_young = False
+        if age is not None and age <= min_wait:
+            log(f"Did not build paper {nr}--recently built.")
+            too_young = True
+        return too_young
+    
+    def check_queue(self, nr):
+        """Check whether the queue currently contains a build request for a PR.
+        
+        Parameters:
+        ------------ 
+        nr: int, the number of the PR to check
+        """
+        in_queue = False
+        if nr in self.dont_build:
+            log(f"Did not queue paper {nr}--already in queue.")
+            in_queue = True
+        return in_queue
+        
+    def check_age_and_queue(self, nr):
+        """Check whether the PR is old enough or whether it is already in queue.
+        
+        Parameters:
+        ------------ 
+        nr: int, the number of the PR to check
+        """
+        return self.check_age(nr) or self.check_queue(nr)
+    
+    def report_status(self, nr):
+        """prints status notification from status_file for paper `nr` 
+        
+        Parameters:
+        ------------ 
+        nr: int, the number of the PR to check
+        """
+        with io.open(status_file(nr), 'r') as f:
+            status = json.load(f)['status']
+
+        if status == 'success':
+            print(f"Completed build for paper {nr}.")
+        else: 
+            print(f"Paper for {nr} did not build successfully.")
+
+
+    async def queue_builder(self, loop=None):
+        """Manage queue and trigger builds, report results.
+        
+        loop: asyncio.loop, the loop on which to be running these tasks
+        """
+        while True:
+            # await an item from the queue
+            nr = await self.queue.get()
+            # launch subprocess to build item
+            with ThreadPoolExecutor(max_workers=1) as e:
```

Comment:
How many CPUs do we have on Heroku?  We could ramp this up, if multiple are available.

Reply:
When i just ran `heroku run grep -c processor /proc/cpuinfo` I got 8 back… but I don't know if that will use up our dyno hours more quickly